cloudigrade/api/clouds/aws/tasks/imageprep.py

```diff
@@ -209,8 +210,13 @@ def copy_ami_snapshot(  # noqa: C901
         create_aws_machine_image_copy(ami_id, reference_ami_id)
         ami_id = reference_ami_id
 
-    # Create volume from snapshot copy
-    create_volume.delay(ami_id, snapshot_copy_id)
+    # We only need the snapshot to launch the inspection now, so...
+    launch_inspection_instance.delay(ami_id, snapshot_copy_id)
+    # Schedule a task to cleanup the snapshot once the inspection is done.
+    delete_snapshot.apply_async(
```

Comment:
I assume the answer is probably "no", but do we get any indication we could process when the instance stops? Using `countdown` like this does looke like a reasonable alternative if the instance's life is a "fire and forget" situation, and this is probably fine with a large enough `INSPECTION_SNAPSHOT_CLEAN_UP_INITIAL_DELAY` value, but I wanted to ask just in case.

Is there anything involving the `DeleteOnTermination` argument or similar that could include and affect our `snapshot_copy_id`? Again, I figure this is probably "no", but I wanted to check.

Reply:
Even if we don't know when an instance stops, what are your thoughts on having `launch_inspection_instance` call `delete_snapshot`? If there's an unexpected backlog of `launch_inspection_instance` tasks to process, that would reduce the possibility that a long-waiting `launch_inspection_instance` doesn't have its corresponding `delete_snapshot` before it starts.

Reply:
Oh… maybe that risk is tempered by the `MachineImage.INSPECTED` check later in `delete_snapshot`? Still curious about your thoughts, though. 🙂

Reply:
So originally my plan was to put the delete call in when we processed the results, then, knowing that the AMI was inspected we would clean up the snapshot. Unfortunately we have no knowledge of what the snapshot_id is at that point, I didn't feel it was important enough to make (likely very many small) changes to pass said value around.

The snapshot is used in the RunInstance call to get it to automagically spawn and attach a volume to the instance on boot, which WILL clean up its own created volume (hence us not needing to create/delete/herd them anymore), but it will not delete the source snapshot. So that was not an option for clean up either.

So what I did is work with the following assumption, when I fire the inspection task, I know that a short time from now the image will be inspected, and the snapshot no longer valuable. So I set a delayed task to run the delete_snapshot task, which I have explicitly modified to verify that the source ami is marked as inspected. Otherwise I assume the snapshot is of value still and keep it around, as delete_snapshot will successfully exit without any deletion after creating a future task to check on it again.

Reply:
I'll acknowledge that there is a missing case of well what if it is in error, or pending state? But I'm not sure if that has to necessarily be solved right now.

Reply:
This confirms my thoughts, and I'm okay with the path you've chosen. Thanks for explaining!